rfcs/20250416-gemm-external-zp-precompute/README.md

```diff
@@ -0,0 +1,87 @@
+# User-Supplied Precomputed Zero Points for GEMM
+
+## Introduction
+
+For modern LLM networks, which usually are a bunch of relatively small GEMMs, performance is priority #1.
+To optimize LLMs for performance (i.e. use DPAS and DP4A, the hardware matmul instructions of Intel GPUs)
+a novel approach has been proposed which is to dynamically downcast both A and B matrices from a floating
+point type to INT8 — but to avoid losing too much precision the downcasting algorithm should also produce
+zero points.
+
+Of particular interest is the variation where the B matrix, or "weights", which is usually constant, gets
+downcast ahead of time. The problem is that zero points on B mean that A (the "sources") is to be reduced
+by the K dimension to multiply the resulting reductions to the ZPs extracted from B; see formulas below.
+
+Calculating partial K reductions of A within the GEMM kernel would never be efficient; there is, however,
+a way to perform them painlessly while downcasting A. Everything that needs adding to GEMM to use them is
+an argument identifier to take that reduction buffer from the user.
```

Comment:
IIUC, B is constant in the case of interest, and user should already have zero-points of B known when computing the reduction of A blocks.
If that is the case, why can't they pre-compute the whole compensation term(`-zp(B)*block_sum(A)`), and pass that instead?

Reply:
Well, they certainly can, but what if the same A is used in multiple GEMMs with different Bs?
The proposed scheme, though, isn't getting in the way if the user decides to precompute everything and just pass a dummy scalar B ZP equal to 1.

Reply:
@petercad Can you please add more details? I think I'm missing something. From the formula `R` is already `M x G` and multiplying it by zero-points + reduction should give a vector of size `M` so it's not a matrix.

![image](https://github.com/user-attachments/assets/954d9d6c-0d9c-4560-a884-f12c216df1c2)

Reply:
@echeresh, what you see here is not a matrix product, but a single element of C (column #m, row #n) being computed.

Reply:
@hidefromkgb Hmm, okay. But I see `m` index being used in the `R`-related term while there is no `n`. I assumed this is a broadcast by `n`, isn't it?

Reply:
@petercad I guess the question is about what are the shapes we are caring about:
- precomputing the full compensation term into bias is M x N elements (and user does not need to pass B zero-points)
- taking col_sum for A is M x (K / G), and zero-points for B is N x (K/G) elements. So a total of (K/G) x (M + N) elements.

So for larger shapes, indeed, the bias term would be bigger. But in that case, I expect we would be compute bound, and the size of the bias should be small anyway compared to A and B matrices. For example if we take something like llama2 shapes like M=N=4096, and K=11008, taking the compensation into bias term directly should be negligible, no?

Reply:
@mgouicem That's true, however for larger shapes you still have the compute overhead -- the amount of precomputation is on the same order as GEMM itself, $\Theta(m\cdot n\cdot k)$.